### PR TITLE
Improve gene search

### PR DESCRIPTION
**Pull request by Kristian-A**
# Description
Load all genes in memory and perform all operations such as searching and sorting on the client's side. This improves the performance as it removes the need to wait for the back-end to fetch the data and send it to the front-end.  

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-2024

#### Link to staging deployment URL (or set N/A)
https://ui-kristian-ui782.scp-staging.biomage.net/

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

**Comment by StefanBabukov** [comment]
Can you make the gene search ignore lowercase/uppercase letters? I want to be able to search for lyz and see Lyz2

File: src/__test__/components/data-exploration/differential-expression-tool/DiffExprResults.test.jsx [hunk 1]
```diff
@@ -241,32 +240,34 @@ describe('DiffExprResults', () => {
 
     // // Wait for side-effect to propagate (properties loading and loaded).
     await waitForActions(withResultStore, [DIFF_EXPR_ORDERING_SET, DIFF_EXPR_LOADING, DIFF_EXPR_LOADED]);
+    const entries = component.find(".ant-table-tbody").children();
 
-    expect(fetchWork).toHaveBeenCalledWith(
```
Comment (StefanBabukov) [inline src/__test__/components/data-exploration/differential-expression-tool/DiffExprResults.test.jsx:245]: fetchwork is still called only once right? to get all the genes? So maybe you should still have a check for it here

File: src/__test__/components/data-exploration/gene-list-tool/FilterGenes.test.jsx [hunk 2]
```diff
@@ -17,36 +16,13 @@ describe('FilterGenes', () => {
...
-    act(() => {
-      search.getElement().props.onSearch('tgf');
-    });
+    const searchBox = component.find('.ant-input');
 
-    expect(mockFilter).toHaveBeenCalledTimes(2);
-    expect(mockFilter).toHaveBeenNthCalledWith(1, { filterOption: 'Starts with', text: null });
-    expect(mockFilter).toHaveBeenNthCalledWith(2, { filterOption: 'Starts with', text: 'tgf' });
+    searchBox.simulate('change', { target: { value: 'A' } });
+    expect(searchBox.instance().value).toContain('A');
```
Comment (StefanBabukov) [inline src/__test__/components/data-exploration/gene-list-tool/FilterGenes.test.jsx]: What is the point of this test? Inputing a letter and checking if its there after?

Comment (Kristian-A) [reply src/__test__/components/data-exploration/gene-list-tool/FilterGenes.test.jsx]: Yes, it was there checking if the correct string was sent to the api. It is in fact pointless now. I will remove it...

Comment (StefanBabukov) [reply src/__test__/components/data-exploration/gene-list-tool/FilterGenes.test.jsx]: maybe you should check if the correct genes are displayed/filtered

Comment (Kristian-A) [reply src/__test__/components/data-exploration/gene-list-tool/FilterGenes.test.jsx]: I have a check for this already

Comment (StefanBabukov) [reply src/__test__/components/data-exploration/gene-list-tool/FilterGenes.test.jsx]: I can see it for differential expression but not for the filter genes? Or is it reusing the same component/function?

File: src/components/data-exploration/generic-gene-table/FilterGenes.jsx [hunk 3]
```diff
@@ -9,6 +9,10 @@ import PropTypes from 'prop-types';
 const { Search } = Input;
 const { Option } = Select;
 
+const sanitizeString = (str) => {
+  console.log(typeof str); return str.replace(/[\s|-]+/g, '')
```
Comment (StefanBabukov) [inline src/components/data-exploration/generic-gene-table/FilterGenes.jsx]: remove log

File: src/components/data-exploration/generic-gene-table/GeneTable.jsx [hunk 4]
```diff
@@ -7,47 +7,70 @@ import _ from 'lodash';
...
 import GeneSelectionStatus from 'redux/actions/genes/geneSelectionStatus';
-import { geneTableUpdateReason } from 'utils/geneTable/geneTableUpdateReason';
 import FocusButton from 'components/FocusButton';
 import PlatformError from 'components/PlatformError';
-import useLazyEffect from 'utils/customHooks/useLazyEffect';
 import GeneSelectionMenu from 'components/data-exploration/generic-gene-table/GeneSelectionMenu';
 import FilterGenes from 'components/data-exploration/generic-gene-table/FilterGenes';
 import Loader from 'components/Loader';
 
+const valueComparator = (key) => (a, b) => {
```
Comment (StefanBabukov) [inline src/components/data-exploration/generic-gene-table/GeneTable.jsx:17]: Can you tell me why you need this function?

File: src/components/data-exploration/generic-gene-table/GeneTable.jsx [hunk 5]
```diff
@@ -73,14 +95,17 @@ const GeneTable = (props) => {
     } else if (filterOption === 'Contains') {
       searchPattern = text;
     }
+    console.log(searchPattern);
```
Comment (StefanBabukov) [inline src/components/data-exploration/generic-gene-table/GeneTable.jsx]: remove log

File: src/components/data-exploration/generic-gene-table/GeneTable.jsx (hunk 4 above)
Comment (Kristian-A) [reply src/components/data-exploration/generic-gene-table/GeneTable.jsx:17]: This is used for sorting the columns. The way react tables sort the values is that you pass in a comparator function to it and upon clicking, items are compared against each other in order to get sorted. 

Here is how the two tables (Gene list and Differential expression) are created with different columns:
```
const newColumns = propColumns.map((column) => {
  const modifiedColumn = { ...column, dataIndex: column.key };

  if (column.sorter) {
    modifiedColumn.sorter = valueComparator(column.key);
    modifiedColumn.sortOrder = getSortOrder(column.key);
  }

  return modifiedColumn;
});
```

As their columns have different types I created a generic comparator which handles both strings and numbers. As the column type is not known at the time of creation this solution made the most sense to me.

File: src/components/data-exploration/generic-gene-table/GeneTable.jsx [hunk 6]
```diff
@@ -7,47 +7,71 @@ import _ from 'lodash';
...
   } = props;
 
   const dispatch = useDispatch();
   const selectedGenes = useSelector((state) => state.genes.selected);
   const [geneNameFilterState, setGeneNameFilterState] = useState({});
 
+  const tableStateAllEntries = {
+    pagination: {
+      current: 1,
+      pageSize: 1000000,
```
Comment (StefanBabukov) [inline src/components/data-exploration/generic-gene-table/GeneTable.jsx]: What is this hardcoded page size 1 million? do we even need it? I think you can use field total here to know the page size, judging by the number of genes (if thats what total shows)?

File: src/__test__/components/data-exploration/gene-list-tool/FilterGenes.test.jsx (hunk 2 above)
Comment (Kristian-A) [reply src/__test__/components/data-exploration/gene-list-tool/FilterGenes.test.jsx]: Yes, they are reusing the same component

**Review by StefanBabukov — APPROVED** [review]
lgtm

**Pull request merged**